## Introduction

This workflow identifies RNA isoforms using either cDNA or direct RNA (dRNA) 
Oxford Nanopore reads.

### Preprocesing
cDNA reads are initially preprocessed by [pychopper](https://github.com/nanoporetech/pychopper) 
for the identification of full-length reads, as well as trimming and orientation correction (This step is omitted for 
 direct RNA reads).

### Reference-aided approach
* Full length reads are mapped to a supplied reference genome using [minimap2](https://github.com/lh3/minimap2)
* Transcripts are assembled by [stringtie](http://ccb.jhu.edu/software/stringtie) 
in long read mode (with or without a guide reference annotation) to generate the GFF annotation.
* The annotation generated by the pipeline is compared to the reference annotation. 
using [gffcompare](http://ccb.jhu.edu/software/stringtie/gffcompare.shtml)

### de novo-based approach (experimental!)
* Sequence clusters are generated using [isONclust2](https://github.com/nanoporetech/isONclust2)
  * If a reference genome is supplied, cluster quality metrics are determined by comparing    
  with clusters generated from a minimap2 alignment.
* A consensus sequence for each cluster is generated using [spoa](https://github.com/rvaser/spoa)
* Three rounds of polishing using racon and minimap2 to give a final polished CDS for each gene.
* Full-length reads are then mapped to these polished CDS.
* Transcripts are assembled by stringtie as for the reference-based approach.
* __Note__: This approach is currently not supported with direct RNA reads.

### Workflow inputs
- Directory containing cDNA/direct RNA reads. Or a directory containing subdirectories each with reads from different samples
  (in fastq/fastq.gz format)
- Reference genome in fasta format (required for reference-based assembly).
- Optional reference annotation in GFF2/3 format.